### DISCUSSION TITLE
Scaling Isso Horizontally with Marmot

### DISCUSSION THREAD

**Discussion by maxpert** [posativ/isso#960] [Show and tell]
# Start with [demo](https://maxpert.github.io/marmot/demo)

## What is Marmot?
Marmot is an distributed SQLite replicator that runs as a side-car to you service, and replicates data across cluster using NATS.

## What is Isso?
Isso is a commenting system like Disqus. Users can edit or delete own comments (within 15 minutes by default). Best part about Isso is that it uses SQLite!

## What is Fly.io?
Fly is a platform for running full stack apps and databases close to your users. Compute jobs at Fly.io are virtualized using Firecracker, the virtualization engine developed at AWS as the engine for Lambda and Fargate.

## Why should I care?
This demo effectively shows how Isso can be pushed closer to the edge. These nodes scale up or down based on traffic, and write from everywhere. This can horizontally scale your Isso close to the user. With NATS embedded into Marmot, a sharded RAFT is used to capture changes, and replay them across the fly nodes (multi-primary replicas).

## [Source code](https://gitlab.com/maxpert/marmot-denokv)

**Comment by ix5** [comment]
That certainly looks interesting. Maybe some performance improvements for Isso could obviate the need for clusters though.

**Reply by maxpert** [reply]
IMO this adds additional value of horizontal scaling. So that if your main node goes down Isso still stays up and running. When the old node comes back all the data will be merged and replayed.

**Comment by jelmer** [comment]
Does this run on a non-proprietary platform, i.e. something that is not fly.io?

**Reply by maxpert** [reply]
Absolutely! Marmot is designed to work platform independently. This is just an example because a lot of developers uses Fly.io and love the serverless behavior. I know people running Marmot on their VPS hosts, and using it as distributed cache.
